src/jit/flowgraph.cpp

```diff
@@ -5746,14 +5746,47 @@ void          Compiler::fgFindBasicBlocks()
 
     if (compIsForInlining())
     {
+        bool hasReturnBlocks = false;
+        bool hasMoreThanOneReturnBlock = false;
+
+        for (BasicBlock* block = fgFirstBB; block != nullptr; block = block->bbNext)
+        {
+            if (block->bbJumpKind == BBJ_RETURN)
+            {
+                if (hasReturnBlocks)
+                {
+                    hasMoreThanOneReturnBlock = true;
+                    break;
+                }
+
+                hasReturnBlocks = true;
+            }
+        }
+
+        if (!hasReturnBlocks)
+        {
+            //
+            // Mark the call node as "no return". The inliner might ignore CALLEE_DOES_NOT_RETURN and
+            // fail inline for a different reasons. In that case we still want to make the "no return"
+            // information available to the caller as it can impact caller's code quality.
+            //
+
+            impInlineInfo->iciCall->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
+
+            compInlineResult->NoteInt(InlineObservation::CALLEE_DOES_NOT_RETURN, static_cast<int>(fgBBcount));
```

Comment:
Yep, I've come to the same conclusion. It seems a bit strange to react to the basic block count observation instead of the no return observation but it works for now. Maybe later we could report the bb count earlier, it seems to be reported unnecessarily late.